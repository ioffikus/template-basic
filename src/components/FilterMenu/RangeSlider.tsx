import React from 'react';
import { Slider } from 'antd';
import { SliderValue } from 'antd/lib/slider';
import _ from 'lodash';

const marks = {
  0: '0',
  250: '250',
  500: '500',
  750: '750',
  1000: '1000',
};

interface IProps {
  value: SliderValue;
  fieldSlug: string;
  valueType: string;
  filterType: string;
  onChange: (value: SliderValue, fieldSlug: string, valueType: string, filterType: string) => void;
}

interface IState {
  value: SliderValue;
}

class RangeSlider extends React.PureComponent<IProps, IState> {
  state: IState = {
    value: this.props.value,
  };
  readonly defaultValue: SliderValue = [0, 100];

  componentWillReceiveProps(nextProps: IProps) {
    if (_.isEmpty(nextProps.value)) {
      this.setState({ value: this.defaultValue });
    }
  }

  handleChange = (value: SliderValue) => {
    this.setState({ value });
  };

  releaseHandler = (value: SliderValue) => {
    this.props.onChange(value, this.props.fieldSlug, this.props.valueType, this.props.filterType);
  };

  render() {
    return (
      <div className="root">
        <Slider
          marks={marks}
          step={50}
          defaultValue={this.defaultValue}
          value={this.state.value}
          range
          max={1000}
          onChange={this.handleChange}
          onAfterChange={this.releaseHandler}
        />
      </div>
    );
  }
}

export default RangeSlider;
