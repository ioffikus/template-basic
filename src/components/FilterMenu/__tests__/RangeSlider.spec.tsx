import React from 'react';
import { shallow } from 'enzyme';
import { Slider } from 'antd';
import filterStateMock from 'src/store/__mocks__/filterStore';
import RangeSlider from '../RangeSlider';

describe(null, () => {
  it('render() correct with initialState', () => {
    const rangeSliderMock = filterStateMock.results[0];
    const { field_slug: fieldSlug, value_type: valueType, filter_type: filterType } = rangeSliderMock;
    const value: [number, number] = null;
    const onChange = jest.fn();
    const wrapper = shallow(
      <RangeSlider
        value={value}
        fieldSlug={fieldSlug}
        valueType={valueType}
        filterType={filterType}
        onChange={onChange}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('render() correct with value is different from initial', () => {
    const rangeSliderMock = filterStateMock.results[0];
    const { field_slug: fieldSlug, value_type: valueType, filter_type: filterType } = rangeSliderMock;
    const value: [number, number] = [50, 100];
    const onChange = jest.fn();
    const wrapper = shallow(
      <RangeSlider
        value={value}
        fieldSlug={fieldSlug}
        valueType={valueType}
        filterType={filterType}
        onChange={onChange}
      />,
    );
    const sliderInstance = wrapper
      .find(Slider)
      .dive()
      .instance() as Slider;
    expect(sliderInstance.props.value).toEqual(value);
    expect(wrapper).toMatchSnapshot();
  });
  it('this.props.onChange()', () => {
    const rangeSliderMock = filterStateMock.results[0];
    const { field_slug: fieldSlug, value_type: valueType, filter_type: filterType } = rangeSliderMock;
    const value: [number, number] = null;
    const newValue: [number, number] = [150, 600];
    const onChange = jest.fn();
    const wrapper = shallow(
      <RangeSlider
        value={value}
        fieldSlug={fieldSlug}
        valueType={valueType}
        filterType={filterType}
        onChange={onChange}
      />,
    );
    const sliderInstance = wrapper
      .find(Slider)
      .dive()
      .instance() as Slider;
    sliderInstance.props.onAfterChange(newValue);
    expect(onChange).toBeCalledWith(newValue, fieldSlug, valueType, filterType);
  });
  it('Slider.onChange()', () => {
    const rangeSliderMock = filterStateMock.results[0];
    const { field_slug: fieldSlug, value_type: valueType, filter_type: filterType } = rangeSliderMock;
    const value: [number, number] = null;
    const newValue: [number, number] = [150, 600];
    const onChange = jest.fn();
    const wrapper = shallow(
      <RangeSlider
        value={value}
        fieldSlug={fieldSlug}
        valueType={valueType}
        filterType={filterType}
        onChange={onChange}
      />,
    );
    const sliderInstanceAfter = wrapper
      .find(Slider)
      .dive()
      .instance() as Slider;
    sliderInstanceAfter.props.onChange(newValue);
    wrapper.update();
    const sliderInstanceBefore = wrapper
      .find(Slider)
      .dive()
      .instance() as Slider;
    expect(sliderInstanceBefore.props.value).toEqual(newValue);
  });
});
