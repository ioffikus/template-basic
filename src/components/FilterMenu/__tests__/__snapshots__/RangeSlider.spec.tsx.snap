// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() correct with initialState 1`] = `
<div
  className="root"
>
  <Slider
    defaultValue={
      Array [
        0,
        100,
      ]
    }
    marks={
      Object {
        "0": "0",
        "1000": "1000",
        "250": "250",
        "500": "500",
        "750": "750",
      }
    }
    max={1000}
    onAfterChange={[Function]}
    onChange={[Function]}
    prefixCls="ant-slider"
    range={true}
    step={50}
    tipFormatter={[Function]}
    tooltipPrefixCls="ant-tooltip"
    value={null}
  />
</div>
`;

exports[`render() correct with value is different from initial 1`] = `
<div
  className="root"
>
  <Slider
    defaultValue={
      Array [
        0,
        100,
      ]
    }
    marks={
      Object {
        "0": "0",
        "1000": "1000",
        "250": "250",
        "500": "500",
        "750": "750",
      }
    }
    max={1000}
    onAfterChange={[Function]}
    onChange={[Function]}
    prefixCls="ant-slider"
    range={true}
    step={50}
    tipFormatter={[Function]}
    tooltipPrefixCls="ant-tooltip"
    value={
      Array [
        50,
        100,
      ]
    }
  />
</div>
`;
